app/gui2/shared/retry.ts

```diff
@@ -49,15 +60,55 @@ export async function exponentialBackoff<T, E>(
       onSuccess(retries)
       return result
     } catch (error) {
-      if (retries >= maxRetries) throw error
+      if (retries >= maxRetries) {
+        onFailure(error as E, retries)
+        throw error
+      }
       if (onBeforeRetry(error as E, retries, maxRetries, delay) === false) throw error
       await wait(delay)
     }
   }
 }
 
-export const onBeforeRetry: (message: string) => BackoffOptions<any>['onBeforeRetry'] =
-  (message) => (error, retryCount, maxRetries, delay) => {
-    console.error(message + ` (${retryCount}/${maxRetries} retries), retrying after ${delay}ms...`)
+export function onBeforeRetry(
+  description: string,
+): NonNullable<BackoffOptions<any>['onBeforeRetry']> {
```

Comment:
I know it's an old code, but I think it's a bit cryptic. I assume this is just a sort of "default" callback? The fact that the name collide with option name does not help. Same for `onFailure` below.

Reply:
this is actually new code, but yeah. i guess i'll rename to `defaultOnBeforeRetry` for now? i don't really have any good ideas for this one